todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/taskdetail/TaskDetailViewModel.kt

```diff
@@ -1,108 +1,111 @@
+import androidx.lifecycle.AndroidViewModel
+import androidx.lifecycle.LiveData
+import androidx.lifecycle.MutableLiveData
+import androidx.lifecycle.Transformations
+import com.example.android.architecture.blueprints.todoapp.Event
 import com.example.android.architecture.blueprints.todoapp.R
-import com.example.android.architecture.blueprints.todoapp.SingleLiveEvent
 import com.example.android.architecture.blueprints.todoapp.data.Task
 import com.example.android.architecture.blueprints.todoapp.data.source.TasksDataSource
 import com.example.android.architecture.blueprints.todoapp.data.source.TasksRepository
-import com.example.android.architecture.blueprints.todoapp.tasks.TasksFragment
-
 
 /**
  * Listens to user actions from the list item in ([TasksFragment]) and redirects them to the
  * Fragment's actions listener.
  */
-class TaskDetailViewModel(
-        context: Application,
-        private val tasksRepository: TasksRepository
-) : AndroidViewModel(context), TasksDataSource.GetTaskCallback {
-
-    val task = ObservableField<Task>()
-    val completed = ObservableBoolean()
-    val editTaskCommand = SingleLiveEvent<Void>()
-    val deleteTaskCommand = SingleLiveEvent<Void>()
-    val snackbarMessage = SingleLiveEvent<Int>()
-    var isDataLoading = false
-        private set
-    val isDataAvailable
-        get() = task.get() != null
+open class TaskDetailViewModel(context: Application, private val tasksRepository: TasksRepository) :
+    AndroidViewModel(context), TasksDataSource.GetTaskCallback {
+
+    private val _task = MutableLiveData<Task>()
+
+    private val _isDataAvailable = MutableLiveData<Boolean>()
+
+    private val _dataLoading = MutableLiveData<Boolean>()
+
+    val editTaskCommand = MutableLiveData<Event<Any>>()
+
+    val deleteTaskCommand = MutableLiveData<Event<Any>>()
+
+    private val _snackbarText = MutableLiveData<Event<Int>>()
+
+    // This LiveData depends on another so we can use a transformation.
+    val completed: LiveData<Boolean> = Transformations.map(_task) {
+            input -> input.isCompleted
+    }
+
+    val snackbarMessage: LiveData<Event<Int>>
+        get() = _snackbarText
+
+    val task: LiveData<Task>
+        get() = _task
+
+    val isDataAvailable: LiveData<Boolean>
+        get() = _isDataAvailable
+
+    val dataLoading: LiveData<Boolean>
+        get() = _dataLoading
+
+    protected val taskId: String?
+        get() = _task.value!!.id
 
     fun deleteTask() {
-        task.get()?.let {
-            tasksRepository.deleteTask(it.id)
-            deleteTaskCommand.call()
+        if (_task.value != null) {
+            tasksRepository.deleteTask(_task.value!!.id)
+            deleteTaskCommand.value = Event(Any())
         }
     }
 
     fun editTask() {
-        editTaskCommand.call()
+        editTaskCommand.value = Event(Any())
     }
 
     fun setCompleted(completed: Boolean) {
-        if (isDataLoading) {
+        if (_dataLoading.value!!) {
```

Comment:
can we avoid `!!` in this method?

Reply:
Missed all the kotlinfier !!s. Refactored the whole class to avoid all of them.